### PR TITLE
Server 162. Closes #162

### LINKED ISSUE #162: Creation of a user in CT triggers also creation of an account in AAD
## Description
As a user of CT I want to have one place where I can create new users for the platform that is consistently represented in the platform services

## Acceptance criteria

- [x] User creation to check for existance of a user with the supplied email address in both CT users table and AAD
- [x] Create new CT user and AAD user if no such user account with the given email address exists
- [x] Create new CT user if the user does not exist in CT but does exist in AAD
- [x] If the user does exist in CT and not in AAD then throw an error
- [x] Return the existing user if the user already exists in both CT + AAD

### PR DESCRIPTION
**Pull request by valentinyanakiev**
### Describe the background of your pull request

Added Microsoft Graph module and logic for AAD user management.

### Additional context
1. In order to call downstream api resource (MS Graph API) we need to replace our Bearer Auth token with another access token using the On-behalf-of flow.
2. In order to obtain the token, client secret is required (passed as env variable **AAD_CLIENT_SECRET** in the application) and also scopes need to be provided - in our case we provide them with **AAD_API_SCOPE** (in .env.default).
3. The **AAD_CLIENT_SECRET** env variable is sensitive and is not added to the .env.default. Contact me (Valentin) to provide you one or create one yourself in AAD.
4. In order to persist the bearer token between requests in-memory cache is used.
5. In order to activate the flow, you need to use the ct admin user (admin@playgroundcherrytwist.onmicrosoft.com) or you need to provide yourself Directory.ReadWrite.All permissions.
6. Isomorphic-fetch is a hard requirement for the ms graph client library.
7. In order to activate OBO flow and use it in ms graph client interface is implemented (the interface is defined in the ms graph library).
8. Some obsolete code was removed.

### Governance 

- [x] Documentation is added
- [ ] Test cases are added or updated
- [x] I've read the contributing document https://github.com/cherrytwist/.github/blob/master/CONTRIBUTING.md
- [x] I've read and understand the Code of Conduct https://github.com/cherrytwist/.github/blob/master/CODE_OF_CONDUCT.md
- [x] I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/cherrytwist/Coordination/blob/master/LICENSE 
     and the contributor license agreement: tba

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,6 +52,7 @@
     "typeorm": "^0.2.28"
   },
   "devDependencies": {
+    "@microsoft/microsoft-graph-types": "^1.23.0",
```
Comment (techsmyth) [inline package.json:53]: why only a dev dependency?

File: README.md [hunk 2]
```diff
@@ -86,6 +86,8 @@ Define AAD_TENANT, AAD_CLIENT environment variables - e.g. locally in .env envir
 ```bash
 AAD_TENANT=[tenant (directory) ID]
 AAD_CLIENT= [client (application) ID]
+AAD_API_SCOPE= [API Scopes Required for Downstream APIs, in our case Microsoft Graph API]
+AAD_CLIENT_SECRET=Client secret obtained from the AAD App Registration
```
Comment (techsmyth) [inline README.md]: Is this not environment specific i.e. that it is the client registration for the playground environment? Am not too worried about this leaking out as it is only our playground setup.

File: package.json (hunk 1 above)
Comment (atstoyanov) [reply package.json:53]: types are development dependency, they are not needed for production. If you review the whole `package.json` you'll see that all types are under devDependecies, except `@types/passport-azure-ad` which have to be moved to devDependecies

File: package.json [hunk 3]
```diff
@@ -24,19 +24,25 @@
...
+    "@microsoft/microsoft-graph-client": "^2.1.0",
     "@nestjs/common": "^7.4.4",
     "@nestjs/config": "^0.5.0",
     "@nestjs/core": "^7.4.4",
     "@nestjs/graphql": "^7.7.0",
     "@nestjs/passport": "^7.1.0",
     "@nestjs/platform-express": "^7.0.0",
     "@nestjs/typeorm": "^7.1.4",
+    "@types/node-fetch": "^2.5.7",
     "@types/passport-azure-ad": "^4.0.7",
```
Comment (atstoyanov) [inline package.json]: Move them into "devDependencies".

Comment (valentinyanakiev) [reply package.json]: Addressed in latest commit.

File: README.md (hunk 2 above)
Comment (valentinyanakiev) [reply README.md]: Addressed in the latest commit.

File: src/domain/user/user.service.ts [hunk 4]
```diff
@@ -53,6 +55,19 @@ export class UserService {
     return this.userRepository.findOne({ email: email });
   }
 
+  async userExistsAadCt(email: string): Promise<boolean> {
```
Comment (techsmyth) [inline src/domain/user/user.service.ts]: Looks like this function is not used - keep it or now? Removing it also removes knowledge of MSGraph from teh User service which is probably correct.

Comment (valentinyanakiev) [reply src/domain/user/user.service.ts]: Addressed in commit c590131c0265b53e16dcf1b2160f43fc4613f547

**Review by techsmyth — APPROVED** [review]

**Pull request merged**